### PR TITLE
Cellular Network C++ Solution Using Sets

### PR DESCRIPTION
**Pull request by maggie-j-liu**
I added a C++ solution using sets to CF Cellular Network. #729

### CODE REVIEW

File: solutions/CF Cellular Network.mdx [hunk 1]
```diff
@@ -2,15 +2,72 @@
...
+
+```cpp
+#include <iostream>
+#include <set>
+#include <algorithm>
+using namespace std;
+
+int main()
+{
+    int n, m;
```
Comment (bqi343) [inline solutions/CF Cellular Network.mdx:28]: convert to tabs as mentioned here: https://usaco.guide/general/contributing

**Review by bqi343 — CHANGES_REQUESTED** [review]

**Review by frodakcin — APPROVED** [review]
Looks good.

File: solutions/CF Cellular Network.mdx [hunk 2]
```diff
@@ -2,62 +2,64 @@
...
+
+<CPPSection>
+
+## Solution 1 - Binary Search on a Sorted Array
 
 ```cpp
 #include <bits/stdc++.h>
 using namespace std;
 
 // returns the first index in the array that is >= value, or arr.size() if no such index exists
```
Comment (bqi343) [inline solutions/CF Cellular Network.mdx:20]: > // returns the first index in the array that is >= value, or arr.size() if no such index exists

this seems wrong ... @thecodingwizard I think you added this?

**Review by bqi343 — CHANGES_REQUESTED** [review]

File: solutions/CF Cellular Network.mdx (hunk 2 above)
Comment (bqi343) [reply solutions/CF Cellular Network.mdx:20]: actually it seems right but confusing ... seems odd to end a loop when `lo` *exceeds* `hi`?

Comment (thecodingwizard) [reply solutions/CF Cellular Network.mdx:20]: D: it highkey took me 30 minutes to write that binary search function; our implementations are highly inconsistent across solutions / guide / slides (ex. https://usaco.guide/silver/binary-search-ans?lang=cpp#introduction)

We should standardize binary search impl; should we just do this?

```cpp
int lo = 0, hi = n-1, ans = [defaultAns];

while (lo < hi) {
    int mid = (lo+hi)/2;
    if (f(mid)) lo = mid + 1;
    else hi = mid;
}
```

That was how I learned it at least :/

Comment (frodakcin) [reply solutions/CF Cellular Network.mdx:20]: My go-to:

```
int l=-1, r=n; // [0, l]: valid answer --- [r, n]: invalid answers
while(l<r)
{
  int m=(l+r)/2;
  if(ok(m)) l=m;
  else r=m;
}
return r;
```

Comment (frodakcin) [reply solutions/CF Cellular Network.mdx:20]: what i wrote ^^^ is equivalent to upper_bound in cpp I believe

Comment (bqi343) [reply solutions/CF Cellular Network.mdx:20]: >  ans = [defaultAns];

why do you need an `ans` variable?

```
int l=-1, r=n; // [0, l]: valid answer --- [r, n]: invalid answers
while(l<r)
{
  int m=(l+r)/2;
  if(ok(m)) l=m;
  else r=m;
}
return r;
```

If the return value is supposed to be `-1` then won't this loop infinitely?

Comment (bqi343) [reply solutions/CF Cellular Network.mdx:20]: > https://usaco.guide/silver/binary-search-ans?lang=cpp#introduction

I think we should just remove the first implementation from here

Comment (thecodingwizard) [reply solutions/CF Cellular Network.mdx:20]: > why do you need an ans variable?

Err I think I typically do `if (ok(mid)) l = mid, ans = mid;` or something like that. It's probably not needed though.

> I think we should just remove the first implementation from here

I think the while loop impl is easier to approach compared to the for loop though?

Comment (frodakcin) [reply solutions/CF Cellular Network.mdx:20]: > If the return value is supposed to be -1 then won't this loop infinitely?

yes oops.
I meant `while(l+1<r)`

**Pull request merged**